### ISSUE TITLE
Loss becomes NAN with larger, more layered network

### ISSUE THREAD

**Issue by Bestehorn** [apache/incubator-mxnet#9478]
## Description
When the number of layers/nodes in gluon.Sequential structures increases, the SoftMax cost function will return NaN values and then prevent learning. See attached jupyter notebook for reproduction:
[mlp-gluon-NAN.ipynb.txt](https://github.com/apache/incubator-mxnet/files/1643935/mlp-gluon-NAN.ipynb.txt)
The above example is from the [MXNET tutorial/book website](http://gluon.mxnet.io/chapter03_deep-neural-networks/mlp-gluon.html#Faster-modeling-with-gluon.nn.Sequential) and has just some modifications on the structure of the network (more layers/nodes).

## Minimum reproducible example
See the file above. Main change compared to the tutorial website (see link above) is here:
```
num_hidden = 512
net = gluon.nn.Sequential()
with net.name_scope():
    net.add(gluon.nn.Dense(num_hidden, activation="relu"))
    net.add(gluon.nn.Dense(num_hidden, activation="relu"))
    net.add(gluon.nn.Dense(num_hidden, activation="relu"))
    net.add(gluon.nn.Dense(num_hidden, activation="relu"))
    net.add(gluon.nn.Dense(num_hidden, activation="relu"))
    net.add(gluon.nn.Dense(num_hidden, activation="relu"))
    net.add(gluon.nn.Dense(num_hidden, activation="relu"))
    net.add(gluon.nn.Dense(num_hidden, activation="relu"))
    net.add(gluon.nn.Dense(num_hidden, activation="relu"))
    net.add(gluon.nn.Dense(num_outputs))
```
This means I just increased the number of nodes and layers.

The output produced by the nootebook is as follows:

> Used 60000 examples in this epoch...
> Epoch 0. Loss: nan, Train_acc 0.0987166666667, Test_acc 0.098
> Used 60000 examples in this epoch...
> Epoch 1. Loss: nan, Train_acc 0.0987166666667, Test_acc 0.098
> Used 60000 examples in this epoch...
> [...]

This basically means the network is not improving or changing at all, as the cost function returns NaN.

## Steps to reproduce
Run the jupyter notebook attached to this ticket.

## What have you tried to solve it?

1. Reducing the number of nodes will solve the problem
2. Reducing the number of layers will solve the problem

While both options may be viable for the sample given here, they are not viable in real-world applications that have complex, large (in terms of features) data sets.

## Comments
I know that running this example with such a large network is an "overkill" that slows down learning etc. I actually came across this problem with a much more complex data set where each sample has 120+k features. Since I cannot share this data set here, I tried to reproduce this problem with data/code that is publicly available.

**Comment by sxjscience** [comment]
Would you also show the initializer and optimizer used to train the network? A good initialization is also essential to avoid exploding relus.

**Comment by Bestehorn** [comment]
Initialization and other stuff has not been changed compared to the tutorial on MXNET [here](http://gluon.mxnet.io/chapter03_deep-neural-networks/mlp-gluon.html#Faster-modeling-with-gluon.nn.Sequential). Here is a copy of the corresponding code from the attached jupyter notebook:
```
net.collect_params().initialize(mx.init.Normal(sigma=.1), ctx=ctx)
softmax_cross_entropy = gluon.loss.SoftmaxCrossEntropyLoss()
trainer = gluon.Trainer(net.collect_params(), 'sgd', {'learning_rate': .01})
```
As stated above, the initialization works fine if the network is "small" in terms of nodes/layers, so I do not think that is really the issue. The input is also normalized during the loading procedure used by the MXNet tutorial by dividing all feature values by 255 (RGB).
```
batch_size = 64
num_inputs = 784
num_outputs = 10
num_examples = 60000
def transform(data, label):
    return **data.astype(np.float32)/255**, label.astype(np.float32)
train_data = mx.gluon.data.DataLoader(mx.gluon.data.vision.MNIST(train=True, transform=transform),
                                      batch_size, shuffle=True)
test_data = mx.gluon.data.DataLoader(mx.gluon.data.vision.MNIST(train=False, transform=transform),
                                     batch_size, shuffle=False)
```

**Comment by Bestehorn** [comment]
First of all: Thanks for your prompt responses and effort to look into this.
While changing sigma to 0.01 has the effect of removing the NaN cost issue, it does not solve the underlying issue that the model is not getting better, i.e., not learning. This is the output I get with sigma=0.01 as instructed by you:

> Epoch 0. Loss: 2.30134551239, Train_acc 0.112366666667, Test_acc 0.1135
> Epoch 1. Loss: 2.30124220098, Train_acc 0.112366666667, Test_acc 0.1135
> Epoch 2. Loss: 2.3012252299, Train_acc 0.112366666667, Test_acc 0.1135
> [...]

Despite having more layers and being slower, the network should still converge to a solution which it apparently does not.

**Comment by sxjscience** [comment]
@Bestehorn I think it's related to the optimizer. You can try to use different learning rate/optimization methods to see the effect. You can try adam, which trains faster
```python
trainer = gluon.Trainer(net.collect_params(), 'adam', {'learning_rate': 1E-3})
```
